core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java

```diff
@@ -457,11 +484,85 @@ private synchronized PrioritizedSplitRunner pollNextSplitWorker()
         return null;
     }
 
+    @VisibleForTesting
+    public static Optional<StuckSplitInterrupter> createStuckSplitInterrupter(
+            boolean enableInterruptStuckSplits,
+            Duration interruptStuckSplitsTimeout,
+            Duration stuckSplitDetectionInterval,
+            Predicate<List<StackTraceElement>> potentiallyStuckSplitPredicate)
+    {
+        if (!enableInterruptStuckSplits) {
+            return Optional.empty();
+        }
+
+        return Optional.of(new StuckSplitInterrupter(interruptStuckSplitsTimeout, stuckSplitDetectionInterval, potentiallyStuckSplitPredicate));
+    }
+
+    private static class StuckSplitInterrupter
+    {
+        private final Duration stuckSplitDetectionInterval;
+        private final Duration interruptStuckSplitsTimeout;
+        private final Predicate<List<StackTraceElement>> potentiallyStuckSplitPredicate;
+
+        public StuckSplitInterrupter(Duration interruptStuckSplitsTimeout, Duration stuckSplitDetectionInterval, Predicate<List<StackTraceElement>> potentiallyStuckSplitPredicate)
+        {
+            requireNonNull(interruptStuckSplitsTimeout, "interruptStuckSplitsTimeout is null");
+            requireNonNull(stuckSplitDetectionInterval, "stuckSplitDetectionInterval is null");
+            requireNonNull(potentiallyStuckSplitPredicate, "potentiallyStuckSplitPredicate is null");
+
+            checkArgument(interruptStuckSplitsTimeout.compareTo(SPLIT_RUN_QUANTA) >= 0, "interruptStuckSplitsTimeout must be at least %s", SPLIT_RUN_QUANTA);
+
+            this.interruptStuckSplitsTimeout = interruptStuckSplitsTimeout;
+            this.stuckSplitDetectionInterval = stuckSplitDetectionInterval;
+            this.potentiallyStuckSplitPredicate = potentiallyStuckSplitPredicate;
+        }
+
+        public Duration getStuckSplitDetectionInterval()
+        {
+            return stuckSplitDetectionInterval;
+        }
+
+        private String toStackString(List<StackTraceElement> stackTraceElements)
+        {
+            String stackString = stackTraceElements.stream()
+                    .map(Object::toString)
+                    .collect(joining(lineSeparator()));
+
+            return stackString;
+        }
+
+        private void interruptStuckSplits(SortedSet<RunningSplitInfo> runningSplitInfos, Ticker ticker)
```

Comment:
is this comment addressed? https://github.com/trinodb/trino/pull/12479#discussion_r881715903

Reply:
I believe not yet.